Alerting: Append org ID to alert notification URLs

neel1996 opened the pull request:
- The orgId field has been appended as query param to the dashboard and panel URL that are sent as part of an alert notification
- Resolves #47424

**What this PR does / why we need it**:

The PR addresses #47424. The dashboard and the panel source URLs that are included in an alert notification currently does not include the orgId reference. If an user is on a different org and tries to land on the dashboard URL included in the alert, then the user will receive an error from UI, as the target dashboard is not within the current org account

![image](https://user-images.githubusercontent.com/47709856/196348101-d0646753-7b83-418a-9e16-8bc4c318c4f9.png)


**Which issue(s) this PR fixes**:

The above issue has been addressed by including the `orgId` as a query param in both the dashboard and the panel URLs that are sent as part of an alert notification.

**After fix**

Below is a sample Discord alert with the updated URLs

![image](https://user-images.githubusercontent.com/47709856/196264246-769581a1-2464-41fd-9440-6fd76daa9bb6.png)

Switched to an org that does not have the source of the alert

![image](https://user-images.githubusercontent.com/47709856/196264657-caf171ee-db56-4443-b5d6-6dca18d180bd.png)

Landing on the dashboard using the URL from the alert

![image](https://user-images.githubusercontent.com/47709856/196264884-f2e32cf8-41d2-4c8c-902f-a221b8aafbb0.png)

Fixes #47424

neel1996 committed:
AlertManager: Append org ID to alert notification URLs
- The orgId field has been appended as query param to the dashboard and panel URL that are sent as part of an alert notification
- Resolves #47424

neel1996 committed:
AlertManager: Append orgId to alert source URL
- Add orgId as query param to the panel URL that raised the alert
- Parse dashboard URL to add orgId as query param instead of query param hardcoding
- Resolved #47424

neel1996 commented:
Hi @alexweav,

This is the PR to address issue #47424.

**Summary of the change:** The `orgId` field has been appended as query param to the URLs sent in the alert notifications which are scoped based on organisations (dashboard, panel & alert source)

joeblubaugh commented on pkg/services/ngalert/notifier/channels/template_data.go:
```diff
@@ -89,6 +89,34 @@ func extendAlert(alert template.Alert, externalURL string, logger log.Logger) *E
+		if len(orgId) > 0 {
+			orgIdParam := "orgId=" + orgId
+
+			dashboardUrl.RawQuery = orgIdParam
```
Why use RawQuery here and Query.Add three lines down? Let's be consistent.

neel1996 replied:
Have extracted the logic of appending the query param to a private function to keep the logic consistent.

joeblubaugh commented on pkg/services/ngalert/notifier/channels/template_data.go:
```diff
@@ -89,6 +89,34 @@ func extendAlert(alert template.Alert, externalURL string, logger log.Logger) *E
 			extended.PanelURL = u.String()
 		}
+
+		generatorUrl, gErr := url.Parse(extended.GeneratorURL)
```
Can you please add a test that validates the orgID is appended when these are set, but the url is otherwise the same?

neel1996 replied:
Hi,

Currently there are two tests covering these scenarios

[// Firing with dashboard and panel ID](https://github.com/grafana/grafana/blob/8911d64574bda76a24b5345d18cb2ba69a483abe/pkg/services/ngalert/notifier/channels/default_template_test.go#L19) - This has the orgId in the annotations and expects the query param to be appended to the generator URL

[// Firing without dashboard and panel ID](https://github.com/grafana/grafana/blob/8911d64574bda76a24b5345d18cb2ba69a483abe/pkg/services/ngalert/notifier/channels/default_template_test.go#L29) - This expects the generated URL without the orgId

Should the above mentioned tests sufice?

neel1996 committed:
Refactor the logic to append orgId query param
- Extracted a common function to add orgId query param to alert notification URLs

neel1996 commented:
Hi @joeblubaugh 

I have pushed the changes based on the review discussions. Please have a look and share your thoughts

grobinson-grafana commented on pkg/services/ngalert/notifier/channels/template_data.go:
```diff
@@ -89,6 +89,25 @@ func extendAlert(alert template.Alert, externalURL string, logger log.Logger) *E
 			extended.PanelURL = u.String()
 		}
+
+		generatorUrl, gErr := url.Parse(extended.GeneratorURL)
```
You can just use `err` here, it's OK to re-use it

```suggestion
		generatorUrl, err := url.Parse(extended.GeneratorURL)
```

neel1996 replied:
Hi, The suggested change has been made to re-use the same `err` variable

grobinson-grafana commented on pkg/services/ngalert/notifier/channels/template_data.go:
```diff
@@ -89,6 +89,25 @@ func extendAlert(alert template.Alert, externalURL string, logger log.Logger) *E
+			return extended
+		}
+
+		dashboardUrl, dErr := url.Parse(extended.DashboardURL)
```
```suggestion
		dashboardUrl, err := url.Parse(extended.DashboardURL)
```

neel1996 committed:
Refactor to re-use same error variable

grobinson-grafana reviewed (approved):
Looks good to me!
